### ISSUE TITLE
Increased intra-node communication latency going from v3 to v4

### ISSUE THREAD

**Issue by pllopis** [open-mpi/ompi#7784]
Hi all,

When moving to version 4 (more precisely, 4.0.3), I have detected an increased latency between processes within a node.

## Symptoms

Using OpenMPI 4.0.3

```
mpirun --report-bindings -mca btl self,vader --map-by socket --bind-to core -n 2 -host hpc002 mpi/pt2pt/osu_latency -m 0

[hpc002.cern.ch:1628113] MCW rank 0 bound to socket 0[core 0[hwt 0-1]]: [BB/../../../../../../..][../../../../../../../..]
[hpc002.cern.ch:1628113] MCW rank 1 bound to socket 1[core 8[hwt 0-1]]: [../../../../../../../..][BB/../../../../../../..]
# OSU MPI Latency Test v5.6.3
# Size          Latency (us)
0                       2.79
```
Using OpenMPI 3.0.0
```
mpirun --report-bindings -mca btl self,vader --map-by socket --bind-to core -n 2 -host hpc002 mpi/pt2pt/osu_latency -m 0

[hpc002.cern.ch:1628258] MCW rank 0 bound to socket 0[core 0[hwt 0-1]]: [BB/../../../../../../..][../../../../../../../..]
[hpc002.cern.ch:1628258] MCW rank 1 bound to socket 1[core 8[hwt 0-1]]: [../../../../../../../..][BB/../../../../../../..]
# OSU MPI Latency Test v5.6.3
# Size          Latency (us)
0                       0.35
```

As you can see I have tried to narrow down the BTLs and process mapping/pinning being used to make everything more explicit. But in spite of both OpenMPI builds behaving seemingly the same way, there's an important difference in inter-process latency.

## Things I've looked at/tried so far

I have also noticed that there are a few btl settings that have changed between 3.00 and 4.0.3 for vader, such as `btl_vader_free_list_max` and `btl_vader_backing_directory`, but playing with them did not yield improved results. I tried setting btl_vader_free_list_max to 4096 and setting btl_vader_backing_directory to /tmp.

If I run the bove commands prefixed with `strace -f -e trace=open`, I see that for 3.0.0 the last open calls were:
```
[pid 1628898] open("/dev/infiniband/rdma_cm", O_RDWR|O_CLOEXEC) = 17
[pid 1628898] open("/dev/infiniband/rdma_cm", O_RDWR|O_CLOEXEC) = 18
[pid 1628898] open("/dev/infiniband/rdma_cm", O_RDWR|O_CLOEXEC) = 18
[pid 1628898] open("/dev/infiniband/rdma_cm", O_RDWR|O_CLOEXEC) = 18
[pid 1628896] open("/tmp/pllopiss/ompi.hpc002.1713620135/jf.18568/1/1/vader_segment.hpc002.1", O_RDWR) = 17
[pid 1628898] open("/tmp/pllopiss/ompi.hpc002.1713620135/jf.18568/1/0/vader_segment.hpc002.0", O_RDWR) = 17
```

While for 4.0.3 they were:
```
[pid 1629025] open("/dev/infiniband/rdma_cm", O_RDWR|O_CLOEXEC) = 24
[pid 1629025] open("/dev/infiniband/rdma_cm", O_RDWR|O_CLOEXEC) = 24
[pid 1629024] open("/dev/infiniband/rdma_cm", O_RDWR|O_CLOEXEC) = 23
[pid 1629024] open("/dev/infiniband/rdma_cm", O_RDWR|O_CLOEXEC) = 24
[pid 1629024] open("/dev/infiniband/rdma_cm", O_RDWR|O_CLOEXEC) = 24
[pid 1629024] open("/dev/infiniband/rdma_cm", O_RDWR|O_CLOEXEC) = 24
```

I was wondering if vader is being used at all, but I can also spot calls to open /dev/shm in the strace for 4.0.3, they're just further up in the log:
```
[pid 1628985] open("/dev/shm/vader_segment.hpc002.48a00001.1", O_RDWR|O_CREAT, 0600) = 15
```

I have attached both traces in case they're helpful (open3 for open calls in OpenMPI 3.0.0, open4 for calls in OpenMPI 4.0.3).
[open3.txt](https://github.com/open-mpi/ompi/files/4730153/open3.txt)
[open4.txt](https://github.com/open-mpi/ompi/files/4730154/open4.txt)

Also, note that I also see the same latency increase (by about ~2 microseconds) with inter-node communications. With OpenMPI 3.0.0 I get barely 3us latency, while with OpenMPI 4.0.3 I see 5.5us latencies. I have focused on a single node for now since I suspect this to be the same issue and to keep debugging simpler.

## Other information about the builds, system, etc.

I've obtained both OpenMPI builds from the tarballs available on the official website. They were both configured and built in the same way:

```
./configure --prefix=%{install_prefix} \
            --libdir=%{install_prefix}/lib64 \
            --enable-mpi-cxx --enable-cxx-exceptions --enable-mpi-thread-multiple --enable-orterun-prefix-by-default \
            --with-slurm --with-pmi=/usr
            CFLAGS='-m64 -O2 -pipe -Wall -Wshadow' \
            CXXFLAGS='-m64 -O2 -pipe -Wall -Weffc++ -Wshadow' \
            FCFLAGS='-m64 -O2 -pipe -Wall' \
            FFLAGS='-m64 -O2 -pipe -Wall'
```

And built using gcc and running on CentOS 7.8 and kernel  3.10.0-957.27.2.el7.x86_64.

Any insights into why latency has suddenly increased and to get the better latency from 3.0.0 would be welcome. Thanks in advance!

Pablo

**Comment by ggouaillardet** [comment]
Pablo,

There is a possibility vader is not used at all. That typically occurs when UCX or MXM or OFI pml or mtl are used.

To force the use of btl/vader, you also need to force pml/ob1

mpirun --mca pml ob1 --mca btl vader,self ...

Cheers,

Gilles

**Comment by pllopis** [comment]
Thanks for the quick reply Gilles,

Indeed, I confirm that I can get sub-microsecond latencies when using pml/ob1. 
My OpenMPI build was built with support for the OFI pml (I haven't tested UCX yet, but I will soon) and I believe was using pml/cm by default.

While I only reported on the pt2pt microbenchmark, the increased latency is present in collective operations as well and is noticeable for small and bigger message sizes (for both local and multi-node).

In light of these findings, I have a few questions:
- Is it possible to use a different pml/mtl combination that provides similar performance to vader? (I'll test with UCX instead of OB1 soon, but any known performance advantages or disadvantages here?)
- Is pml/ob1 gonna be deprecated, or is it just no longer the default for other reasons? While I could just switch the defaults to keep using pml/ob1 (since those are the ones that give the lowest latency for node-local workloads), I'd like to know if sticking to ob1 is a good idea.
- Related to the last question, and possibly answering my own question: I've read that the openib btl is now deprecated, since UCX is now the preferred pml. I also want to run on clusters with iWARP and with Infiniband, but workloads use a combination of e.g. Infiniband with shared memory communication. Ideally I'd keep using a component that is supported (e.g. UCX over Infiniband) but doesn't add extra overhead for shared memory communication that I've experienced. Which means, if I understood well, that I can't use OB1/vader with UCX? (and consequently, the answer to the previous question is "no, sticking to ob1 is not a good idea"?).

Basically, I'd like to get with OpenMPI 4.x equal or better performance than what I was getting with OpenMPI 3/OB1 (ideally using fully supported components, of course).

Thanks!
Pablo

**Comment by ggouaillardet** [comment]
Pablo,

OFI is a mtl component, that means it is used with pml/cm.
UCX is both available as a pml and a btl (btl/uct).
pml/ob1 is not gonna be deprecated, it allows the use of different btl on a
per peer basis
(e.g. btl/vader for intra node communications, and btl/tcp for inter node
communications).
There is no such possibility with the other pml.

I do not use OFI, but IIRC, you must use the 'shm' provider in order to get
the best intra node performances.
Otherwise you would use the loopback interface which is likely less
efficient.

IIRC, btl/openib is now deprecated for Infiniband only, and is still
supported for iWARP.

If you are not happy with the pml/ucx performances for intra node
communications, you can force pml/ob1.
Hopefully, it will use btl/vader for intra node communications, and btl/uct
otherwise.

Cheers,

Gilles

**Comment by pllopis** [comment]
Thanks for the clarification Gilles. I am closing this issue since the question has been resolved. Thanks again!
